### PR TITLE
Refactor new meta chain

### LINKED ISSUE #100: Refactor newMetachain
- Move constants EPOCH_LENGTH from Axiom to Consensus contract. Add a constant MAX_STATE_ROOTS in Consensus contract 

Refactor:Axiom::newMetachain

- Remove all params from axiom:newMetachain()
- Remove all the logic related to maxStateRoot and rootRlpBlockHeader.
- Call consensus.newMetaChain.
- Get the metachainId as the return value of consensus.newMetaChain
- Emit an event MetachainCreated

Refactor: Consensus::newMetachain

- Add a new mapping consensusGateways (metachainId bytes32 -> consensusGateway address)
- Call Axiom::deployMetachainProxies
- Generate metachainId.
- ~~Update the CoreLifetime status to Creation~~ (done in newCore already)
- Update assignments mapping.
- Update consensusGateways mapping
- Return metachainId

### PR DESCRIPTION
**Pull request by 0xsarvesh**
fixes #100

### CODE REVIEW

File: contracts/consensus/Consensus.sol [hunk 1]
```diff
@@ -608,46 +636,56 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
+        bytes memory coreSetupCallData = coreSetupData(
+            metachainId_,
+            EPOCH_LENGTH,
             uint256(0), // metablock height
             bytes32(0), // parent hash
             gasTargetDelta, // gas target
             uint256(0), // dynasty
             uint256(0), // accumulated gas
-            _rootBlockHeight
+            0 // todo should be fixed with remove source block height in core-setup.
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol]: Source block height will not be removed.
It was a source block hash, and its already removed.

File: contracts/consensus/Consensus.sol [hunk 2]
```diff
@@ -608,46 +636,56 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
+
+        (address core, address consensusGateway) =
+            axiom.deployMetachainProxies(
+                coreSetupCallData,
+                consensusGatewaySetupCallData
+            );
+
+        require(
+            assignments[metachainId_] == address(0),
+            "A core is already assigned to this metachain."
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol]: We are deploying anchor contract in this function, and then generating the `metachainId`. 
I am not sure if this condition is required.

File: contracts/consensus/Consensus.sol [hunk 3]
```diff
@@ -864,9 +902,10 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
      * @param _dynasty Committed dynasty number.
      * @param _accumulatedGas Accumulated gas.
      * @param _sourceBlockHeight Source block height.
+     *
      * returns Deployed core contract address.
      */
-    function newCore(
+    function coreSetupData(
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol:959]: Maybe we still need `newCore` function. When the core will halt, a new core will be deployed for the same `metachainId`

File: contracts/axiom/Axiom.sol [hunk 4]
```diff
@@ -27,11 +27,12 @@ contract Axiom is AxiomI, ProxyFactory, ConsensusModule {
 
     using SafeMath for uint256;
```
Comment (deepesh-kn) [inline contracts/axiom/Axiom.sol:29]: Add a new line.

File: contracts/axiom/Axiom.sol [hunk 5]
```diff
@@ -27,11 +27,12 @@ contract Axiom is AxiomI, ProxyFactory, ConsensusModule {
...
 
+    /* Events */
 
-    /* Constants */
+    /* Emitted when new metachain is created */
+    event MetachainCreated(bytes32 metachainId);
 
-    /** Epoch length */
-    uint256 public constant EPOCH_LENGTH = uint256(100);
+    /* Constants */
```
Comment (deepesh-kn) [inline contracts/axiom/Axiom.sol:37]: Add a new line.

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
Looks good. 👍 
Just a few minor comments.

File: contracts/consensus/Consensus.sol (hunk 2 above)
Comment (0xsarvesh) [reply contracts/consensus/Consensus.sol]: I agree with the argument, however, anchor deployment happens in another contract. 
We have followed pattern previously to not trust other contract implementation and write validations explicitly.

File: contracts/consensus/Consensus.sol (hunk 3 above)
Comment (0xsarvesh) [reply contracts/consensus/Consensus.sol:959]: Option 1:
In that case, we need to decouple core deployment from `deployMetachainProxies`? 
Or we leave it for now and make a decision when we have more clarity. 

Option2: 
We can have public separate functions for the deployment of each contract and also have `deployMetaChainProxies` which stitches them.

Comment (0xsarvesh) [reply contracts/consensus/Consensus.sol:959]: Based on the discussion, we have an external function in`Axiom.newCore`.  There is no action item on this point.

File: contracts/consensus/Consensus.sol (hunk 2 above)
Comment (0xsarvesh) [reply contracts/consensus/Consensus.sol]: Discussion conclusion:  Removed the require with the argument that we trust the axiom contract.

**Review by deepesh-kn — APPROVED** [review]
LGTM :shipit:

**Pull request merged**